src/streamlink/stream/hls.py

```diff
@@ -232,21 +232,24 @@ def reload_playlist(self):
         sequences = [Sequence(media_sequence + i, s)
                      for i, s in enumerate(playlist.segments)]
 
-        if sequences:
-            self.process_sequences(playlist, sequences)
+        self.process_sequences(playlist, sequences)
 
     def _set_playlist_reload_time(self, playlist, sequences):
         self.playlist_reload_time = (playlist.target_duration
-                                     or sequences[-1].segment.duration)
+                                     or len(sequences) > 0 and sequences[-1].segment.duration)
 
     def process_sequences(self, playlist, sequences):
+        self._set_playlist_reload_time(playlist, sequences)
+
+        if not sequences:
+            return
+
```

Comment:
I'm not sure if this could potentially break other HLS streams. @back-to @beardypig ?

The reason I've changed this here is that the default playlist reload time is 15 and if there are no segments in the first playlist "refresh" (first download), then the _set_playlist_reload_time method does not get called. This makes the Twitch preroll ad test wait unnecessarily long for 15 seconds.

(sorry about posting the same comment twice)

Reply:
seems reasonable, I don't think it would break anything

Reply:
Ok, good.

What about the rest of the changes? There's a bit of refactoring here which I haven't really mentioned yet. Eg. I've removed the stream dependency on the M3U8Parser, removed the now unnecessary method decorator and also removed the custom StreamWriter class, as the ad filtering is now done by the parser which simply ignores the next line if the tag above is identified as an ad. This further means that there are no log messages anymore which indicate the start or end of an ad break, but that couldn't be done anyway due to the lack of proper annotations in the playlist (previously we had the start and end tags which made that possible). Basically the same issue with just relying on the discontinuity tags, which you've tried last year when they implemented the embedded ads system the first time.